### PR TITLE
removed globals, added postprocessing

### PR DESCRIPTION
**Pull request by lumpy-nerdwallet**
there are still 2 c files we need to go over; let's do that tomorrow

### CODE REVIEW

File: src/generate_features_member.lua [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+       u'aage', u'cl',
+       u'pastdue', u'lastpaid',
+       u'pmtfrac', u'late_30', u'late_60', u'late_90']]--
+local assertx = require 'assertx'
+local luatz = require 'luatz.init'
+local MAX_TS = 2147483648
+local open_mort_codes = {['SM']=true, ['CV']=true, ['FR']=true, ['VM']=true, ['FL']=true, ['RM']=true, ['CMG']=true}
+
+
+local function validate_input(args)
```
Comment (itsindra) [inline src/generate_features_member.lua]: you are validating the numeric aspect, not validating input so change things accordingly. Also if you want we can have a common validators lib

File: src/generate_features_member.lua [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+	local ub = args['ub'] or math.maxinteger
+	assertx(var >= lb and var <= ub, "Variable ", tostring(var_nm), " with value ", var, " is out of range, which is [", lb, ", ", ub, "].")
+	return var
+end
+
+
+local function generate_tradelines_features(tradelines, ts) 
+	-- verify above that time will be verified before putting in here
+	-- Each tradeline is a dictionary; we are interested in many things.
+	local q, ty, o, derog, oage, aage, cl, pastdue, lastpaid, pmtfrac, late_30, late_60, late_90 = 0, 0, 0, 0, 0, 0, 0, 0, math.maxinteger, 0, 0, 0, 0
```
Comment (itsindra) [inline src/generate_features_member.lua]: very hardto see what is what, break if more than 3 or 4

File: src/generate_features_member.lua [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+	local sum_age, mly_pmt, sum_past_due = 0, 0, 0
+	local open_mort = false
+	for _, tl in pairs(tradelines) do
+		assert(tl['open_closed_symbol'], 'Open-closed symbol not present.')
+		assert(tl['account_subtype'], 'Account subtype not present.')
+		if tl['account_subtype'] == 'Collection Account' then
```
Comment (itsindra) [inline src/generate_features_member.lua]: too many lines in one function, function should be at most 20 -30 lines

File: src/get_num_features.lua [hunk 4]
```diff
@@ -17,7 +17,8 @@ local function get_nested_length(table)
 end
 
 local function get_num_features()
-  assert(g_dt_feature, 'g_dt_feature not loaded.')
+  local g_dt_feature = assert(cache.get("g_dt_feature"),
```
Comment (itsindra) [inline src/get_num_features.lua]: remove g_ as g_ stands for globals and there are no globals in our lua space

File: src/get_referrer_type_cd.lua [hunk 5]
```diff
@@ -7,15 +7,20 @@ local function get_referrer_type_cd(args)
...
-	local table_rd_search = assert(cache.get("table_rd_search"), "Cache missing table_rd_search.")
-	]]--
+	local g_table_isn = assert(cache.get("g_table_isn"),
+		"Cache missing g_table_isn.")
+	local g_table_mvc = assert(cache.get("g_table_mvc"),
+		"Cache missing g_table_mvc.")
+	local g_table_rd_sm = assert(cache.get("g_table_rd_sm"),
+		"Cache missing g_table_rd_sm.")
+	local g_table_rd_search = assert(cache.get("g_table_rd_search"),
+		"Cache missing g_table_rd_search.")
```
Comment (itsindra) [inline src/get_referrer_type_cd.lua]: If a comment is unnecessary remove it. , also this function is too long

File: src/one_hot_encoding.lua [hunk 6]
```diff
@@ -17,7 +18,8 @@ local function one_hot_encoding(feature_table)
 	2. This will suddenly fail if any of the non-categorical variables are not numeric.
 	3. If there is a value within the categorical variables that do not exist in ENCODING_RULES, none of the values will be encoded.
 	]]--
-	assert(g_dt_feature, "g_dt_feature not loaded.")
+	local g_dt_feature = assert(cache.get("g_dt_feature"),
```
Comment (itsindra) [inline src/one_hot_encoding.lua]: need to review this with you offline, dont understand your elseif block

File: src/post_proc_preds.lua [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+local function post_proc_preds(
+  opvec, -- feature vector, now a float* of n_opvec length
+  n_opvec, -- number of elements in feature vector. should be 21
+  out_buf, -- output buffer for debugging 
+  sz_out_buf -- size of above
+  )
+  opvec = ffi.cast("float*", opvec)
+  local out_features = {}
+  -- read the g_ccid_mapping
+  local g_ccid_mapping = assert(cache.get('g_ccid_mapping'),
```
Comment (itsindra) [inline src/post_proc_preds.lua]: remove g_

File: src/post_proc_preds.lua [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    'G_CCID_MAPPING missing from cache.')
+  local n_cards = assert(get_n_cards(), 'get_n_cards failed.')
+  assertx(n_opvec == n_cards, 'feature vector length is ,',
+    tostring(n_opvec), ', but g_ccid_mapping says length is ', 
+    tostring(n_cards))
+  for idx, ccid in pairs(g_ccid_mapping) do
+    local ccid = assert(tonumber(ccid), 'g_ccid_mapping faulty')
```
Comment (itsindra) [inline src/post_proc_preds.lua]: use different name than ccid in the loop or inside

File: src/time_to_vars.lua [hunk 9]
```diff
@@ -35,6 +35,11 @@ local function time_to_vars(t, tz)
 	local local_t = tz_value:localise(t) -- local_t is a timestamp
 	local local_tt = luatz.timetable.new_from_timestamp(local_t) -- local_tt is a table
 	local utc_offset = tz_value:find_current(NY_2018).gmtoff / 3600 
+	if utc_offset == math.floor(utc_offset) and string.sub(tostring(utc_offset), -2) ~= '.0' then -- if it's an int, and not int.0
+		utc_offset = tostring(utc_offset) .. '.0'
+	else
+		utc_offset = tostring(utc_offset)
```
Comment (itsindra) [inline src/time_to_vars.lua]: extremely large function

File: test_rts_lua/valid_config.lua [hunk 10]
```diff
@@ -143,6 +143,10 @@ return [=[{
       "COMMENT" : " XXXX ",
       "VALUE"  : "/opt/ab/dt_feature.lua"
     },
+    "CCID_MAPPING" : { 
+      "COMMENT" : " XXXX ",
```
Comment (itsindra) [inline test_rts_lua/valid_config.lua]: keep valid comments, initially things were being hacked together

File: test_visitor_transforms/test_generate_features.lua [hunk 11]
```diff
@@ -1,16 +1,38 @@
...
 local test_data = 'generate_features_data.json'
-_G.g_table_isn, _G.g_table_mvc, _G.g_table_rd_sm, _G.g_table_rd_search = assert(dofile('../config/referrer_class_file.lua'))
-_G.g_postal_cd_features = assert(dofile('../config/postal_cd_features.lua'))
---[[cache.put("table_postal_cd_features", table_postal_cd_features)
-cache.put("table_isn", table_isn)
-cache.put("table_mvc", table_mvc)
-cache.put("table_rd_sm", table_rd_sm)
-cache.put("table_rd_search", table_rd_search)
+--_G.g_table_isn, _G.g_table_mvc, _G.g_table_rd_sm, _G.g_table_rd_search = assert(dofile('../config/referrer_class_file.lua'))
+--_G.g_postal_cd_features = assert(dofile('../config/postal_cd_features.lua'))
```
Comment (itsindra) [inline test_visitor_transforms/test_generate_features.lua]: remove these comments

File: test_visitor_transforms/test_generate_features.lua [hunk 12]
```diff
@@ -1,16 +1,38 @@
...
+local status, g_postal_cd_features = pcall(dofile, '/opt/ab/postal_cd_features.lua')
+assert(status, "loading postal_cd_features.lua failed")
+assert(g_postal_cd_features, "loading g_postal_cd_features failed")
```
Comment (itsindra) [inline test_visitor_transforms/test_generate_features.lua]: i see most of my concerns are repeating. Please fix all instances of the concerns and then i can look at this better

File: src/generate_features_member.lua (hunk 1 above)
Comment (lumpy-nerdwallet) [reply src/generate_features_member.lua]: i called it validate_numeric_input now, i think it's clearer?

File: src/get_referrer_type_cd.lua (hunk 5 above)
Comment (lumpy-nerdwallet) [reply src/get_referrer_type_cd.lua]: done!

File: src/get_num_features.lua (hunk 4 above)
Comment (lumpy-nerdwallet) [reply src/get_num_features.lua]: done

File: src/dt.lua [hunk 13]
```diff
@@ -0,0 +1,27 @@
+local load_cfg = require 'load_config'
+
+function load_config(...)
+  load_cfg.load_transform_features(...)
+end
+
+local xxx = require 'make_feature_vector'
```
Comment (itsindra) [inline src/dt.lua]: dont keep declaring a variable xxx and working with that. Do a better job, refer to ab.lua

File: src/get_n_cards.lua [hunk 14]
```diff
@@ -0,0 +1,13 @@
+local cache = require 'cache'
+
+local function get_n_cards()
+  local ccid_mapping = assert(cache.get("ccid_mapping"),
+  	"CCID_MAPPING not in cache")
+  assert(ccid_mapping, 'ccid_mapping not loaded.')
```
Comment (itsindra) [inline src/get_n_cards.lua]: redundant libe

File: src/get_n_cards.lua [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+  assert(ccid_mapping[0], 'ccid_mapping must be 0-indexed')
+  local count = 0
+  for k, v in pairs(ccid_mapping) do count = count + 1 end
```
Comment (itsindra) [inline src/get_n_cards.lua]: single level or do you need to account for hierarchy?

File: src/get_referrer_type_cd.lua [hunk 16]
```diff
@@ -1,52 +1,73 @@
...
+		"Cache missing table_isn.")
+	local table_rd_sm = assert(cache.get("table_rd_sm"),
+		"Cache missing table_rd_sm.")
+	local table_rd_search = assert(cache.get("table_rd_search"),
+		"Cache missing table_rd_search.")
 	local ext_dw_referral_sk = 0
-	if g_table_isn[domain] ~= nil then
+	if table_isn[domain] ~= nil then
 		ext_dw_referral_sk = 2
 	elseif domain == nil or domain == '' then
```
Comment (itsindra) [inline src/get_referrer_type_cd.lua]: shouldnt you be checking for the domain being nil first?

File: src/one_hot_encoding.lua [hunk 17]
```diff
@@ -1,52 +1,76 @@
...
 			final_value = 1
-		elseif type(cat_encoding_rules) == 'number' then
-			index = cat_encoding_rules
-			if type(raw_value) == 'boolean' then
-				if raw_value == true then
-					final_value = 1
-				else
-					final_value = 0
-				end
+		else -- if it's not {vk:vv} then it must be v
```
Comment (itsindra) [inline src/one_hot_encoding.lua]: update comment

File: test_generator/json_generator.lua [hunk 18]
```diff
@@ -0,0 +1,203 @@
+package.path=package.path .. ";./../src/?.lua"
+local dbg = require 'debugger'
+require 'str'
+math.randomseed(os.time())
+
```
Comment (itsindra) [inline test_generator/json_generator.lua]: name file as test_*.lua
also avoid a generic name like json_geerator, this is not generating json for any table

File: src/get_n_cards.lua (hunk 14 above)
Comment (lumpy-nerdwallet) [reply src/get_n_cards.lua]: done

File: src/dt.lua (hunk 13 above)
Comment (lumpy-nerdwallet) [reply src/dt.lua]: done

File: src/get_n_cards.lua (hunk 15 above)
Comment (lumpy-nerdwallet) [reply src/get_n_cards.lua]: i don't, but it's 0 indexed so i did this. I could also use #

File: src/get_referrer_type_cd.lua (hunk 16 above)
Comment (lumpy-nerdwallet) [reply src/get_referrer_type_cd.lua]: line 53 of the main fn does that

File: src/one_hot_encoding.lua (hunk 17 above)
Comment (lumpy-nerdwallet) [reply src/one_hot_encoding.lua]: done

**Pull request merged**